Framework/Algorithms/src/CylinderAbsorption.cpp

```diff
@@ -45,39 +47,79 @@ void CylinderAbsorption::defineProperties() {
+    return; // not a cylinder
+
+  // get to the underlying ShapeInfo object
+  const auto csgshape = dynamic_cast<const CSGObject *>(&sampleShape);
+  if (!csgshape)
+    return;
+  const auto &shapeObj = csgshape->shapeInfo();
+
+  if (updateRadius)
+    m_cylRadius = shapeObj.radius();
+  if (updateHeight)
+    m_cylHeight = shapeObj.height();
+}
+
 /// Fetch the properties and set the appropriate member variables
 void CylinderAbsorption::retrieveProperties() {
+  m_numSlices = getProperty("NumberOfSlices");
+  m_numAnnuli = getProperty("NumberOfAnnuli");
+
+  bool userSuppliedHeight = false;
+  bool userSuppliedRadius = false;
+
   m_cylHeight = getProperty("CylinderSampleHeight"); // in cm
+  if (!isEmpty(m_cylHeight)) {
+    m_cylHeight *= 0.01; // now in m
+    userSuppliedHeight = true;
+  }
   m_cylRadius = getProperty("CylinderSampleRadius"); // in cm
-  m_cylHeight *= 0.01;                               // now in m
-  m_cylRadius *= 0.01;                               // now in m
+  if (!isEmpty(m_cylRadius)) {
+    m_cylRadius *= 0.01; // now in m
+    userSuppliedRadius = true;
+  }
 
-  m_numSlices = getProperty("NumberOfSlices");
-  m_sliceThickness = m_cylHeight / m_numSlices;
-  m_numAnnuli = getProperty("NumberOfAnnuli");
-  m_deltaR = m_cylRadius / m_numAnnuli;
-
-  /* The number of volume elements is
-   * numslices*(1+2+3+.....+numAnnuli)*6
-   * Since the first annulus is separated in 6 segments, the next one in 12 and
-   * so on.....
-   */
-  m_numVolumeElements = m_numSlices * m_numAnnuli * (m_numAnnuli + 1) * 3;
-
-  if (m_numVolumeElements == 0) {
-    g_log.error() << "Input properties lead to no defined volume elements.\n";
-    throw std::runtime_error("No volume elements defined.");
+  // this declares that at least part of the built-in sample geometry should be
+  // ignored and use the supplied parameters instead
+  m_useSampleShape = !(userSuppliedHeight || userSuppliedRadius);
+
+  // if the user supplied both, then just ignore the built-in shape
+  if (userSuppliedHeight && userSuppliedRadius) {
+    g_log.information(
+        "Chosing user supplied sample geometry in CylinderAbsorption");
+    return;
   }
 
-  m_sampleVolume = m_cylHeight * M_PI * m_cylRadius * m_cylRadius;
+  // get the missing parameters from the sample shape
+  const auto &sampleShape = m_inputWS->sample().getShape();
+  getShapeFromSample(sampleShape, !userSuppliedHeight, !userSuppliedRadius);
 
-  if (m_sampleVolume == 0.0) {
-    g_log.error() << "Defined sample has zero volume.\n";
-    throw std::runtime_error("Sample with zero volume defined.");
+  g_log.information() << "Creating cylinder with radius=" << m_cylRadius
+                      << "m, height=" << m_cylHeight << "m\n";
+  if (isEmpty(m_cylHeight) || isEmpty(m_cylRadius)) {
+    throw std::invalid_argument(
+        "Need to specify both height and radius of cylinder");
   }
+  if (m_cylHeight <= 0. || m_cylRadius <= 0.)
```

Comment:
The values are now optional in the algorithm to allow reading them off of the sample. There is a bug in this code that needs to be fixed (again). When converting from cm to m, the new value no longer passes `isEmpty` correctly. I'll move to using `userSuppliedHeight` and `userSuppliedRadius` which will clarify things.

Reply:
Looking at this a little deeper, the code to create geometry doesn't error check supplied parameters. It would be easy to have a sample with negative radius.

Reply:
Ok, then better leave the check as-is.